File: java/SimpleAk/src/test/java/PongTest.java
/**
 * Created by cqshinn on 2/12/16.
 */
import actor.PongActor;
import akka.actor.ActorRef;
import akka.actor.ActorSystem;
import akka.actor.Props;
import org.junit.Test;
import scala.concurrent.Future;

import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.TimeUnit;

import static scala.compat.java8.FutureConverters.*;
import static akka.pattern.Patterns.ask;
import static akka.pattern.Patterns.pipe;

public class PongTest {
    ActorSystem sActor = ActorSystem.create();
    ActorRef ref = sActor.actorOf(Props.create(PongActor.class));
    public CompletionStage<String> askPong(String message){
        Future sFuture = ask(ref, "Ping", 1000);
        CompletionStage<String> cs = toJava(sFuture);
        return cs;
    }
    @Test
    public void PingTest() throws Exception{

//        Future sFuture = ask(ref, "Ping", 1000);
        final CompletableFuture<String> jFuture = (CompletableFuture<String>) askPong("Ping");
        assert(jFuture.get(1000, TimeUnit.MICROSECONDS).equals("Pong"));

    }

    @Test(expected = ExecutionException.class)
    public void ErrorTest() throws Exception{
        final CompletableFuture<String> jFuture = (CompletableFuture<String>) askPong("unknown");
        jFuture.get(1000, TimeUnit.MILLISECONDS);
        askPong("cause error")
                .handle((res, ex) -> {
                    return ex == null ? CompletableFuture.completedFuture(res) : askPong("Ping");
                }).thenCompose( x -> x );
    }


}

File: scala/SimpleAk/src/test/scala-2.11/PongTest.scala
/**
 * Created by cqshinn on 2/12/16.
 */

import akka.actor.{ActorSystem, Actor}
import akka.pattern.ask
import akka.util.Timeout
import org.scalatest.{Matchers, FunSpecLike}
import scala.concurrent.{Future, Await}
import scala.concurrent.duration._
import scala.language.postfixOps

class PongTest extends FunSpecLike with Matchers {
  import actor.PongActor._
  import scala.concurrent.ExecutionContext.Implicits.global

  val system = ActorSystem()
  implicit val timeout = Timeout(5 seconds)
  val pong = system.actorOf(props)
  def askPong(message: String): Future[String] = (pong ? message).mapTo[String]
  val futureFuture: Future[Future[String]] = askPong("Ping").map(x => { askPong(x) })
  describe("test pong"){
    it("should be response with ping"){
      askPong("Pong").onSuccess({
        case result: String => println("replied with: " + result)
      })
      Thread.sleep(1000)
//      val result = Await.result(future.mapTo[String], 1 second)
//      assert( result == "Ping")
    }

    it("should fail when send a unknown message to actor"){
      val future = pong ? "_"
      intercept[Exception]{
        Await.result(future.mapTo[String], 1 second)
      }
      askPong("causeError 1").onFailure{
        case e: Exception => println("Got exception")
      }
      askPong("causeError 2").recoverWith({
        case ex: Exception => askPong("Pong")
      });
    }
  }

}
